src/reader/mod.rs

```diff
@@ -28,24 +28,43 @@ pub enum ColorOutput {
     Indexed = 1,
 }
 
-#[derive(Debug)]
+#[derive(Clone, Debug)]
 /// Memory limit in bytes. `MemoryLimit::Some(0)` means
 /// that there is no memory limit set.
 pub struct MemoryLimit(pub u32);
 
+impl MemoryLimit {
+    fn buffer_size(&self, color: ColorOutput, width: u16, height: u16) -> Option<usize> {
+        let pixels = u32::from(width) * u32::from(height);
+
+        let bytes_per_pixel = match color {
+            ColorOutput::Indexed => 1,
+            ColorOutput::RGBA => 4,
+        };
+
+        if self.0 > 0 && pixels > self.0 / bytes_per_pixel {
+            None
+        } else {
+            Some(pixels as usize * bytes_per_pixel as usize)
+        }
+    }
+}
+
 /// Options for opening a GIF decoder.
 #[derive(Clone, Debug)]
 pub struct DecodeOptions {
-    memory_limit: u32,
+    memory_limit: MemoryLimit,
     color_output: ColorOutput,
+    check_frame_consistency: bool,
 }
 
 impl DecodeOptions {
     /// Creates a new decoder builder
     pub fn new() -> DecodeOptions {
         DecodeOptions {
-            memory_limit: 50_000_000, // 50 MB
+            memory_limit: MemoryLimit(50_000_000), // 50 MB
             color_output: ColorOutput::Indexed,
+            check_frame_consistency: true,
```

Comment:
Open question: Which is the better default? `false` restores the behaviour pre-`0.17` while `true` is arguably safer for consumers and follow the specification more closely.

Reply:
Let's default to `false` here. This gives the freedom to choose any of the other styles, in particuar the decoder should not crash on it and the risk is limited to unexpected indices (but we still always do allocation checks).